mozart++/core/_event.hpp

```diff
@@ -0,0 +1,255 @@
+            /**
+             * using static function to replace virtual function to avoid unnecessary vtable overhead
+             */
+            template <typename T>
+            static void *create_event(T &&func)
+            {
+                function_type *data = new function_type(std::forward<T>(func));
+                return reinterpret_cast<void *>(data);
+            }
+            static void destroy_event(void *data) noexcept
+            {
+                delete reinterpret_cast<function_type *>(data);
+            }
+            static bool call_on(void *func, void **data) noexcept
+            {
+                static typename make_sequence<sizeof...(ArgsT)>::result seq;
+                return call_impl(func, data, seq);
+            }
+        };
+        /**
+         * Main Class
+         */
+        class event_emit {
+            template <typename...>
+            struct argument_carrier {};
+            struct event {
+                void *data;
+                std::vector<std::type_index> types;
+                void (*destroy_event)(void *) noexcept;
+                bool (*call_on)(void *, void **) noexcept;
+                template <typename T, typename... ArgsT>
+                event(T &&func, argument_carrier<ArgsT...>)
+                {
+                    using impl = event_impl<ArgsT...>;
+                    data = impl::create_event(std::forward<T>(func));
+                    convert_typeinfo<ArgsT...>::convert(types);
+                    destroy_event = &impl::destroy_event;
+                    call_on = &impl::call_on;
+                }
+                ~event()
+                {
+                    destroy_event(data);
+                }
+            };
+            std::unordered_map<std::string, std::forward_list<event>> m_events;
+            template <typename R, typename... ArgsT>
+            void on_impl(const std::string &name,
+                         const std::function<R(ArgsT...)> &func)
+            {
+                m_events[name].emplace_front(func, argument_carrier<ArgsT...>());
+            }
+
+        public:
+            /**
+             * Register an event with handler.
+             * @tparam handler type of the handler
+             * @param name event name
+             * @param handler event handler
+             */
+            template <typename T>
+            void on(const std::string &name, T &&listener)
+            {
+                /**
+                 * Check through type traits
+                 * 1. Listener must be a function
+                 * 2. Listener must returns bool, true represents process finish
+                 */
+                static_assert(!std::is_function<T>::value, "Event must be function");
+                static_assert(std::is_same<bool, parse_function_return_t<T>>::value,
+                              "Function must returns bool");
+                on_impl(name, parse_function<T>(std::forward<T>(listener)));
```

Comment:
EventEmitter is a `poster-subscriber` model, where subscriber should not affect the poster.